File: src/main/java/com.banepig.dcf4j/CommandDispatcher.java
package com.banepig.dcf4j;

import com.banepig.dcf4j.stringcaster.CastType;
import com.banepig.dcf4j.stringcaster.StringCaster;
import sx.blah.discord.api.IDiscordClient;
import sx.blah.discord.api.events.EventDispatcher;
import sx.blah.discord.api.events.EventSubscriber;
import sx.blah.discord.handle.impl.events.guild.channel.message.MessageReceivedEvent;
import sx.blah.discord.handle.obj.IMessage;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

public class CommandDispatcher {
    private IDiscordClient client;
    private StringCaster stringCaster;
    private HashMap<String, ArrayList<CommandExecutor>> registeredCommands = new HashMap<>();
    private String usageMessage = "Invalid usage! Usage: %USAGE%";

    /**
     * Initializes a CommandDispatcher object
     *
     * @param client The client this object should use to listen for events.
     */
    public CommandDispatcher(IDiscordClient client) {
        this.client = client;
        this.stringCaster = new StringCaster(client);

        EventDispatcher dispatcher = client.getDispatcher();
        dispatcher.registerListener(this);
    }

    /**
     * Initializes a CommandDispatcher object
     *
     * @param client The client this object should use to listen for events.
     * @param usageMessage The custom invalid usage message which dcf4j will send to users if they use a command incorrectly.
     */
    public CommandDispatcher(IDiscordClient client, String usageMessage) {
        this.client = client;
        this.stringCaster = new StringCaster(client);

        this.usageMessage = usageMessage;

        EventDispatcher dispatcher = client.getDispatcher();
        dispatcher.registerListener(this);
    }

    /**
     * Registers commands, and methods with the "Commands" annotation will attempt to be called.
     *
     * @param commands The commands to register.
     */
    public void registerCommands(Object... commands) throws IllegalArgumentException {
        for (Object commandClassInstance : commands) {
            Class commandClass = commandClassInstance.getClass();
            for (Method commandExecutorMethod : commandClass.getMethods()) {
                Command annotation = commandExecutorMethod.getAnnotation(Command.class);
                if (annotation == null) continue;
                String label = annotation.label().toLowerCase();
                CommandExecutor commandExecutor = new CommandExecutor(commandClassInstance, commandExecutorMethod);
                registeredCommands.computeIfAbsent(label, k -> new ArrayList<>());
                ArrayList<CommandExecutor> commandsWithLabel = registeredCommands.get(label);
                commandsWithLabel.add(commandExecutor);
                if (!commandExecutor.isAnnotationsValid()) {
                    throw new IllegalAnnotationException();

                }
                registeredCommands.put(label.toLowerCase(), commandsWithLabel);
            }
        }
    }

    /**
     * Unregisters commands, they will no longer be called.
     *
     * @param commands The commands to unregister.
     */
    public void unregisterCommands(Object... commands) {
        for (Object commandClassInstance : commands) {
            Class commandClass = commandClassInstance.getClass();
            for (Method commandExecutorMethod : commandClass.getMethods()) {
                Command annotation = commandExecutorMethod.getAnnotation(Command.class);
                if (annotation == null) continue;
                String label = annotation.label();
                CommandExecutor commandExecutor = new CommandExecutor(commandClassInstance, commandExecutorMethod);
                registeredCommands.computeIfAbsent(label, k -> new ArrayList<>());
                ArrayList<CommandExecutor> commandsWithLabel = registeredCommands.get(label);
                commandsWithLabel.remove(commandExecutor);
                registeredCommands.put(label, commandsWithLabel);
            }
        }
    }

    /**
     * Searches for commands with messages.
     *
     * @param event The message.
     */
    @EventSubscriber
    public void onMessageEvent(MessageReceivedEvent event) {
        List<String> args = Arrays.asList(event.getMessage().getContent().split(" "));

        String label = args.get(0).toLowerCase();
        args = args.subList(1, args.size());
        dispatchCommand(label, args.toArray(new String[args.size()]), event.getMessage());
    }

    /**
     * @param label The label of the command.
     * @param args  The command arguments.
     */
    private void dispatchCommand(String label, String[] args, IMessage message) {
        List<CommandExecutor> commandExecutors = registeredCommands.get(label);
        if (commandExecutors == null || commandExecutors.isEmpty()) return;

        Boolean commandExecuted = false;

        invokeLoop:
        for (CommandExecutor commandExecutor : commandExecutors) {
            Class<?>[] commandExecutorParams = commandExecutor.getCommandExecutor().getParameterTypes();
            Object[] castedArgs = new Object[commandExecutorParams.length];
            castedArgs[0] = message;
            int index = 1;
            for (String argument : args) {
                if (commandExecutorParams.length <= index) continue invokeLoop;
                CastType goalType = StringCaster.getCastType(commandExecutorParams[index]);
                castedArgs[index] = stringCaster.autoCast(goalType, argument);
                index++;
            }
            for (int i = args.length + 1; i < castedArgs.length - args.length; i++) castedArgs[i] = null;
            boolean isCorrectUsage = commandExecutor.tryInvoke(castedArgs);
            if (isCorrectUsage) commandExecuted = true;
        }

        if (!commandExecuted) {
            Command command = commandExecutors.get(0).getAnnotation();
            message.getChannel().sendMessage(usageMessage.replace("%USAGE%", command.usage()));
        }
    }

    public String getUsageMessage() {
        return usageMessage;
    }

    public void setUsageMessage(String usageMessage) {
        this.usageMessage = usageMessage;
    }
}

File: src/main/kotlin/com/banepig/dcf4j/CommandDispatcher.kt
package com.banepig.dcf4j

import sx.blah.discord.api.IDiscordClient
import sx.blah.discord.api.events.EventSubscriber
import sx.blah.discord.handle.impl.events.guild.channel.message.MessageReceivedEvent
import sx.blah.discord.handle.obj.IMessage

class CommandDispatcher(val client: IDiscordClient) {
    init {
        val dispatcher = client.dispatcher
        dispatcher.registerListener(this)
    }

    private val registeredCommands = HashMap<String /*Label*/, ArrayList<CommandHandler> /*CommandMethod*/>()

    /**
     * Registers all commands in a class.
     */
    fun registerCommand(commandInstance: Any) {
        for (method in commandInstance::class.java.methods) {
            if (method.getAnnotationsByType(Command::class.java).isNotEmpty()) registerCommand(CommandHandler(method, commandInstance))
        }
    }

    /**
     * Registers a command by method.
     */
    fun registerCommand(commandHandler: CommandHandler) {
        val commandAnnotation = commandHandler.method.getAnnotation(Command::class.java)
                ?: throw Exception("Command annotation not found")
        val label = commandAnnotation.label.toLowerCase()
        registeredCommands[label] = (registeredCommands[label] ?: ArrayList()).apply { add(commandHandler) }
    }

    @EventSubscriber
    fun onMessage(event: MessageReceivedEvent) {
        val args = ArrayList(event.message.content.split(" "))
        if (args.isEmpty()) return
        val label = args[0].toLowerCase()
        args.removeAt(0)

        var commandSuccessful = false

        for (commandHandler in registeredCommands[label] ?: return) {
            val commandMethod = commandHandler.method
            val parameterTypes = commandMethod.parameterTypes
            if (parameterTypes.isEmpty() || parameterTypes[0] != IMessage::class.java) throw Exception("Commands first parameter is not of IMessage type")
            val castedArgs = ArrayList<Any?>()
            castedArgs.add(event.message)
            for (arg in args.withIndex()) {
                if (parameterTypes.size < arg.index + 1) return
                castedArgs.add(arg.value.to(parameterTypes[arg.index + 1], client))
            }
            val wasDispatched = CommandInvoker.tryDispatch(commandHandler, castedArgs)
            if (wasDispatched) commandSuccessful = true
        }

        if (registeredCommands[label]?.isNotEmpty() == true && !commandSuccessful) {
            event.channel.sendMessage(registeredCommands[label]!![0].method.getAnnotation(Command::class.java).usage)
        }
    }
}